src/web/tokenserver.rs

```diff
@@ -1,26 +1,173 @@
+
+#[derive(Debug)]
+enum MyError {
+    EnvError(env::VarError),
+}
+
+impl From<env::VarError> for MyError {
+    fn from(error: env::VarError) -> Self {
+        MyError::EnvError(error)
+    }
+}
+
+#[derive(Debug, QueryableByName)]
+struct TokenserverUser {
+    #[sql_type = "Bigint"]
+    uid: i64,
+    // This is no longer used. Was for making more than just sync tokens.
+    #[sql_type = "Text"]
+    pattern: String,
+    #[sql_type = "Text"]
+    email: String,
+    #[sql_type = "Bigint"]
+    generation: i64,
+    #[sql_type = "Text"]
+    client_state: String,
+    #[sql_type = "Bigint"]
+    created_at: i64,
+    #[sql_type = "Nullable<Bigint>"]
+    replaced_at: Option<i64>,
+    #[sql_type = "Text"]
+    node: String,
+    #[sql_type = "Nullable<Bigint>"]
+    keys_changed_at: Option<i64>,
+}
+
+#[derive(serde::Serialize)]
+pub struct TokenServerResult {
+    id: String,
+    key: String,
+    uid: String,
+    api_endpoint: String,
+    duration: String,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub struct Claims {
+    pub sub: String,
+    pub iat: i64,
+    pub exp: i64,
+}
 
 pub fn get(
-    request: TokenServerRequest,
+    auth: BearerAuth,
 ) -> impl Future<Output = Result<HttpResponse, BlockingError<ApiError>>> {
-    block(move || get_sync(request).map_err(Into::into)).map_ok(move |_result| {
-        // TODO turn _result into a json response.
+    block(move || get_sync(&auth).map_err(Into::into)).map_ok(move |result| {
         HttpResponse::Ok()
             .content_type("application/json")
-            .body("{}")
+            .body(serde_json::to_string(&result).unwrap())
     })
 }
 
-pub fn get_sync(_request: TokenServerRequest) -> Result<TokenServerResult, ApiError> {
-    // TODO Perform any blocking calls needed to respond to the tokenserver request.
-    Ok(TokenServerResult {})
+pub fn get_sync(auth: &BearerAuth) -> Result<TokenServerResult, ApiError> {
+    // the public rsa components come from
+    // https://oauth.accounts.firefox.com/v1/jwks
+    // TODO we should fetch it from an environment var instead of hardcoding it.
+    let token_data = decode::<Claims>(
+        &auth.token(),
+        &DecodingKey::from_rsa_components("2lDphW0lNZ4w1m9CfmIhC1AxYG9iwihxBdQZo7_6e0TBAi8_TNaoHHI90G9n5d8BQQnNcF4j2vOs006zlXcqGrP27b49KkN3FmbcOMovvfesMseghaqXqqFLALL9us3Wstt_fV_qV7ceRcJq5Hd_Mq85qUgYSfb9qp0vyePb26KEGy4cwO7c9nCna1a_i5rzUEJu6bAtcLS5obSvmsOOpTLHXojKKOnC4LRC3osdR6AU6v3UObKgJlkk_-8LmPhQZqOXiI_TdBpNiw6G_-eishg8V_poPlAnLNd8mfZBam-_7CdUS4-YoOvJZfYjIoboOuVmUrBjogFyDo72EPTReQ", "AQAB"),
+        &Validation::new(Algorithm::RS256),
+    ).map_err(|ee| {
+        ApiError::from(ApiErrorKind::Internal(format!("Unable to decode token_data: {:}", ee)))
+    })?;
+    let email = format!("{:}@api.accounts.firefox.com", token_data.claims.sub);
+
+    // TODO pull out of settings instead
+    let shared_secret = env::var("SYNC_MASTER_SECRET").expect("SYNC_MASTER_SECRET must be set");
```

Comment:
nit: Tempted to move these up a bit in a future release. They'd be the trigger that @pjenvey is looking for and would save the cycles of building the `token_data` above.